### PR TITLE
add SimpleProfiler

### PR DESCRIPTION
**Pull request by llin17**
#134 Add SimpleProfiler to modelkit

### CODE REVIEW

**Comment by CyrilLeMat** [comment]
Great job @llin17 !
I think you should add some tests + some documentation (why not using a profiling at the end of the tutorial ;-) ).

File: modelkit/core/profilers/base.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    @abstractmethod
+    def start(self, model_name: str) -> None:
+        """Define how to start recording an action."""
+
+    @abstractmethod
+    def end(self, model_name: str) -> None:
+        """Define how to record the cost when an action is completed."""
+
+    def summary(self, *args, **kwargs) -> str:
```
Comment (ldeflandre) [inline modelkit/core/profilers/base.py:22]: I am wondering about the wording here.

Why a profiler should implement a `summary` method ?

A profiler need to give a way to query the metrics.

I was looking at https://pypi.org/project/prometheus-client/ or https://docs.datadoghq.com/fr/developers/dogstatsd/?tab=python but they both handle the exposure of metrics automatically.

Probably we should let each profiler implementation the choice on how they expose the metrics. ie. do not define this method in baseclass.
What do you think ?

File: modelkit/core/profilers/simple.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
+import time
+from collections import defaultdict
+from typing import Dict, Union, Set
+import numpy as np
+import pandas as pd
+
+from modelkit.core.profilers.base import BaseProfiler
+
+
+class SimpleProfiler(BaseProfiler):
```
Comment (ldeflandre) [inline modelkit/core/profilers/simple.py:10]: An idea for future PR, we may propose an simple implementation using https://pypi.org/project/prometheus-client/

**Review by ldeflandre — COMMENTED** [review]
Great Job Lu.

File: modelkit/core/profilers/base.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    @contextmanager
+    def profile(self, model_name: str) -> Generator:
+        """
+        Usage:
+            with self.profile('do something'):
+                # do something
+        """
+        try:
+            if model_name == self.main_model_name:
+                self.start_time = time.monotonic()
```
[the remark this answers is not in the document]
Comment (llin17) [reply modelkit/core/profilers/base.py]: Thanks for the comment. Problem resolved with new commits.

File: modelkit/core/profilers/base.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+from abc import ABC, abstractmethod
+from contextlib import contextmanager
+from typing import Generator, Dict, Set
+import time
+
+class BaseProfiler(ABC):
+    """Base Profiler class to be inherited for custom profiler."""
+
+    def __init__(self, model) -> None:
```
[the remark this answers is not in the document]
Comment (llin17) [reply modelkit/core/profilers/base.py]: Thank for the comment. Problem resolved.

**Pull request merged**